core/src/main/java/org/apache/calcite/sql2rel/RelDecorrelator.java

```diff
@@ -1242,6 +1244,37 @@ public Frame decorrelateRel(Join rel) {
     return register(rel, newJoin, mapOldToNewOutputs, corDefOutputs);
   }
 
+  public Frame decorrelateRel(LogicalUnion rel) {
+    return decorrelateRel((Union) rel);
+  }
+
+  public Frame decorrelateRel(Union rel) {
+    assert !cm.mapRefRelToCorRef.containsKey(rel);
+
+    final List<RelNode> oldInputs = rel.getInputs();
+
+    final Frame frame0 = getInvoke(oldInputs.get(0), rel);
+    if (frame0 == null) {
+      return null;
+    }
+
+    relBuilder.push(frame0.r);
+
+    for (int i = 1; i < oldInputs.size(); i++) {
+      final Frame frame = getInvoke(oldInputs.get(i), rel);
+      if (frame == null) {
+        return null;
+      }
+      if (!frame.oldToNewOutputs.equals(frame0.oldToNewOutputs)
+          || !frame.corDefOutputs.equals(frame0.corDefOutputs)) {
```

Comment:
Can you explain more about this check ?

Reply:
Sure, updated.

Reply:
Is it covered by the test cases?

Reply:
@hsyuan Good question ~
`testDecorrelateUnion1` is testing the second point "2. The input frames have the same correlation references and corresponding outputs."

But for the first point (" 1. The input frames have the same mapping of old outputs to new outputs"), I cannot find a way to test.

In current implementation of decorrelation for operators like Project/Filter/Join/Correlate/Union/Aggregate, I didn't find one which permute the mapping in `oldToNewOutputs `, i.e. basically the mapping is derived from parent to child.

For short, in my understanding, Union should always have child frames with the same `oldToNewOutputs`.

Please correct me if I'm wrong.

Reply:
testDecorrelateUnion1 and testDecorrelateUnion2 is for such non-matching case.

Reply:
In my understanding, If the inputs do not have the same mapping, it's hard to derive the oldToNewOutputs for the frame of Union

Reply:
Also when oldToNewOutputs are different between the inputs, I think we should not union the new plan together.